src/plugins/rhbz.c

```diff
@@ -874,3 +890,164 @@ xmlrpc_value *rhbz_search_duphash(struct abrt_xmlrpc *ax,
 
     return bugs;
 }
+
+xmlrpc_value *rhbz_get_sub_components(struct abrt_xmlrpc *ax, const char *product, const char *component)
+{
+    xmlrpc_env env;
+    xmlrpc_env_init(&env);
+
+    /* Product info is stored in a struct as follows:
+     * {"products":({"id":i,"name":s,"description":s,"is_active":b,"default_milestone":s,"default_release":s,"has_unconfirmed":b,"classification":s,"components":({"id":i,"name":s,"description":s,"default_assigned_to":s,"default_qa_contact":s,"default_docs_contact":s,"default_cc":(s),"sort_key":i,"is_active":b,"sub_components":({})})})}
+     *
+     * where:
+     * () = array
+     * {} = struct
+     * b  = bool
+     * i  = int
+     * s  = string
+     * 
+     * The structs in the "sub_components" array are almost identical to the ones in the parent
+     * "components" array. The same is repeated recursively for subcomponents that themselves contain
+     * subcomponents.
+     * 
+     * http://xmlrpc-c.sourceforge.net/doc/libxmlrpc.html#formatstring
+     * 
+     * First, obtain a struct containing one item called "products" whose value is an array containing one struct
+     * containing information about the product we specified in the function arg. The "include_fields" arg
+     * specifies the struct as the following subset of the above:
+     * {"products":({"name":s,"components":({"name":s,"sub_components":({})})})}
+     * 
+     * The "components" struct doesn't have to be specified by itself because we specify its members.
```

Comment:
I’m really struggling to understand this. Specified where? You’re looking at products here and accessing their components.

Reply:
Specified in the "include_fields" array that you're passing as an arg to `abrt_xmlrpc_call`.  
If you do `"{s:(s)}", "include_fields", "components.name"`, you get the product struct conaining the `components` struct, which cointains the `name`. I.e. you don't have to specify that you want the `components` struct because it's already mentioned in `components.name`. However, if you attempt the same with the `sub_components` struct (i.e. `{s:(ss)}, "include_fields", "components.name", "components.sub_components.name"`) you don't get the `sub_components` struct at all (or you get it empty, I don't remember). You have to explicitly include `"components.sub_components"` _and_ `"components.sub_components.name"` as args to the format string.  
  
Anyway, I removed most of the long comments with the struct description.

Reply:
Aha! This is interesting.

https://bugzilla.redhat.com/docs/en/html/integrating/api/Bugzilla/WebService/Product.html


    sub_components

    array of hashes. This value must be specified in the include_fields or extra_fields argument. This feature is a Red Hat customisation.

Maybe that’s a clue.

Reply:
In any case, can you replace the text with this dumbed-down version? I’ll be scratching my head sometime in the future when I read this again.

Reply:
Rewritten, hopefully for the better.